erts/emulator/beam/erl_bif_lists.c

```diff
@@ -203,6 +236,14 @@ static BIF_RETTYPE lists_append_onheap(Process *p, ErtsAppendContext *context) {
 }
 
 static int append_continue(Process *p, ErtsAppendContext *context) {
+    /* Fast-lane when the rhs is nil: return lhs. */
+    if (is_nil(context->rhs_original)) {
+        context->result = context->lhs_original;
+        context->result_cdr = &context->rhs_original;
```

Comment:
We don't need to set these every time we're returning from a yield. I think this part should be moved to `append_start`.

Reply:
I have this to satisfy the assertion here: https://github.com/erlang/otp/blob/695c729be543e5cc01bcf0a13d6e94936b66b357/erts/emulator/beam/erl_bif_lists.c#L259

Maybe that assertion shouldn't apply when we skip appending? We could add a branch above that assertion to check

```c
if (is_nil(context->rhs_original)) {
    BIF_RET(context->lhs_original);
}
```

instead. Another alternative could be to only set these (`context->result` and `context->result_cdr`) when `append_is_proper_list` returns `1`. What do you think?

Reply:
Right, the address can change when trapping. Let's set them when returning `1`, and skip setting `result_cdr` outside of debug builds.